Port the last user of static_analysist

martin-cs opened the pull request:
This ports `value_set_analysis`, which was the last user of deprecated `static_analysist` to use `ait`.  Note that this is *not* the value set analysis code used by `cbmc`; this one is only used by `goto-instrument` (see https://github.com/diffblue/cbmc/issues/6469 ).  There are a few supporting changes to the `ait` infrastructure which are necessary but technically independent.  It is best to review commit by commit.  Note that the test case is added before the porting and works both before and after so most of the output is preserved (the afterwards, `ait` output has slightly more as it also outputs the instructions).

Commit:
Port value_set_analysist and value_set_domain to the ai framework

This removes the last in-tree use of static_analysis.{h,cpp}.
It also gives the option of using value_set_domain with the history
options to perform per-path value-set analysis.

Commit:
Make per-function XML and JSON outputs of ai_baset public

The corresponding text function has been public for a long time and there
is not reason why the JSON and XML ones can't be as they are const and are
called from other public methods.

Commit:
Set the default implementation of make_entry to make_top

This is used by the majority of domains and if you don't want to / can't
implement make_top() then you probably want a custom make_entry() anyway.
This is a net saving of code!

Commit:
Add a domain factory that passes the location of the domain

Commit:
Add a test for the value set analysis in goto-instrument

Previously there were no direct checks for this functionality, only for
users such as the memory model functionality.

Commit:
Adapt accessor methods to ait's newer style

shared_ptr's are returned rather than references as it allows the abstract
interpreter to synthesise domains at query time, either for unexplored
locations or merging multiple traces.

peterschrammel commented on regression/goto-instrument/show-value-sets/main.c:
```diff
@@ -0,0 +1,41 @@
+  int a;
+  u = &a;
+
+  // Test merging and the loss of corelation
```
correlation

martin-cs replied:
Fixed.

martin-cs commented:
At this stage I haven't removed `static_analysist` as I don't know if there are out-of-tree users but @tautschnig ported the other users 8 years ago ( https://github.com/diffblue/cbmc/blob/8a720e04f7706a16390abc8d70b627a74435f1e2/src/analyses/static_analysis.h#L16 ) so maybe it is time to get rid of it?

peterschrammel commented:
> maybe it is time to get rid of it?

no objections

martin-cs commented:
I am seeing unit test failures in https://github.com/diffblue/cbmc/blame/develop/jbmc/unit/pointer-analysis/custom_value_set_analysis.cpp which I think are because `ait` lazily initializes domains while `static_analysist` would initialize everything up front.  This should be fixable but the existence and way the test is written makes me wonder if this is checking an interface that Diffblue code is using, any thoughts @smowton @chrisr-diffblue @peterschrammel @NlightNFotis ?

tautschnig commented on regression/goto-instrument/show-value-sets/test.desc:
```diff
@@ -0,0 +1,27 @@
+but as domains are stored at the start of instructions the following merge will
+over-write it on one branch or the other.  To keep the test portable it is best
+avoided.
+
+
```
Nit pick: unnecessary blank lines?

martin-cs replied:
Fixed

tautschnig commented on src/pointer-analysis/value_set_domain.h:
```diff
@@ -12,64 +12,125 @@ Author: Daniel Kroening, kroening@kroening.com
   {
-    return value_set.make_union(other.value_set);
+    reachable = false;
+    value_set.clear(); //???
```
?

martin-cs replied:
The underlying `value_sett` doesn't really have a clear idea of `top` and `bottom` because it's not really used / formalised like that.  The constructor of `value_set_domain_templatet` would always `value_set.clear()` which means I think it is the right thing to set when `make_bottom`.

tautschnig replied:
Perhaps replace the question marks by exactly that explanation?

martin-cs replied:
Done.

tautschnig commented on src/pointer-analysis/value_set_domain.h:
```diff
@@ -12,64 +12,125 @@ Author: Daniel Kroening, kroening@kroening.com
     value_set.get_reference_set(expr, dest, ns);
   }
+
+  exprt get_return_lhs(locationt to) const
+  {
+    // get predecessor of "to"
+    to--;
```
I am wondering whether `get_return_lhs` should have a different interface: decrementing `to` is an unchecked operation, and would be undefined when `to` already is the beginning of a container.

martin-cs replied:
Yes; it is a risky bit of code and far from an ideal design.  It is moved over because I was trying to preserve functionality and structure where I could.  I should refactor this though.

martin-cs replied:
I have had a look at this and when doing a "return" edge from `END_FUNCTION`, decrementing `to` really is the best way of getting the function call instruction.  That is probably a wider API failure.  However the only case in which this is actually needed is to handle the return value.  So if we are happy return removal has always been done first then it is fine to just remove all of this code.

martin-cs replied:
I am mildly concerned that some of the more obscure users might not ensure this so I think I will just simplify but preserve the functionality.

tautschnig commented on src/pointer-analysis/value_set_domain.h:
```diff
@@ -12,64 +12,125 @@ Author: Daniel Kroening, kroening@kroening.com
+    // get predecessor of "to"
+    to--;
+
+    if(to->is_end_function())
+      return static_cast<const exprt &>(get_nil_irep());
```
Is that really a possible scenario, i.e., can decrementing `to` result in reaching an `END_FUNCTION` instruction? If so, is passing a one-past-the-end iterator to `get_return_lhs` a realistic scenario?

martin-cs replied:
I really don't think it should be.  I think this bit of code is only used in `END_FUNCTION` where `to` is the instruction after the call.

tautschnig commented on src/pointer-analysis/value_set_domain.h:
```diff
@@ -116,4 +177,8 @@ void value_set_domain_templatet<VST>::transform(
 }
 
+// To pass the correct location to the constructor we need a factory
+typedef ai_domain_factory_location_constructort<value_set_domaint>
+  value_set_domain_factoryt;
```
I think we prefer `using` over `typedef` nowadays.

martin-cs replied:
Changed.

smowton commented:
Memories of this far too foggy to be authoritative, but I think this was added so the security project (RIP) could do value-set-analysis-with-tweaks.

peterschrammel commented:
@martin-cs, please adapt the test for use with `ait` instead of `static_analysist`. Any custom value set analysis using this will need to upgrade to `ait` anyway.

martin-cs commented:
@chrisr-diffblue thanks, that does ring a bell...  I will assume that changes to this are not going to be high-risk for Diffblue.

@peterschrammel Will do, I think the changes should be very minimal.

NlightNFotis commented:
A couple of jobs failed because the reference to file has not been removed from the `makefiles.

martin-cs committed:
Remove the static_analysist framework.

This has been deprecated for many many years and only had one user
which was a more obscure feature of goto-instrument.  ait is better
in almost every direction.

martin-cs commented:
@peterschrammel @tautschnig : you both reviewed and approved this back in the distant past.  I am trying to get it moving again.  The PR is essentially identical but with the fix for the issue with the custom value set analysis and removal of static_analysist.  These were the outstanding issues from the reviews.  If you want another look then do shout, otherwise I will merge it when I can get CI to pass!

martin-cs committed:
Tweak headers to keep include-what-you-use happy

martin-cs commented:
I'm going to ignore the patch coverage metric because of the increase of project coverage.
